### PR TITLE
Allow for samplers to be seedable and reproducable

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR ensures that `DataLoaders` made through `Accelerate` are reproducible on distributed systems by allowing `set_epoch` to work on `RandomSamplers` (what gets selected if `shuffle=True` on a `DataLoader`).

Fixes # (issue)

fixes https://github.com/huggingface/transformers/issues/26541
fixes https://github.com/huggingface/accelerate/issues/2016

To test this works, I ran the following script:

```python
from accelerate import Accelerator
from accelerate.utils import set_seed

set_seed(42)
accelerator = Accelerator()
dataset = range(24)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=8, shuffle=True)
dataloader = accelerator.prepare(dataloader)

dataloader.set_epoch(0)
original_outputs = []

for x in dataloader:
    original_outputs += x

original_outputs = torch.stack(original_outputs)

dataloader.set_epoch(0)
new_outputs = []

for x in dataloader:
    new_outputs += x
    
new_outputs = torch.stack(new_outputs)

print(f'Original: {original_outputs}\n')
print(f'New: {new_outputs}\n')
```

When comparing the outputs, these two line up perfectly, ensuring we can in-fact recreate what was originally setup.

For testing we will be relying on the transformers integration CI (see the test runners on main that include the failing test). Linked with https://github.com/huggingface/transformers/pull/26850

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan @LysandreJik

### CODE REVIEW

**Comment by muellerzr** [comment]
Note from @BenjaminBossan:

Eventually we may want to consider whether a user passes in their own `generator` into the sampler init of their own `RandomSampler`, and if we should use that as part of the `__iter__`, such as:

```python
seed = self.epoch + getattr(self.generator, "initial_seed", 0)
```

File: src/accelerate/data_loader.py [hunk 1]
```diff
@@ -64,6 +64,40 @@
...
+        self.epoch = 0
+
+    def __iter__(self):
+        g = torch.Generator()
+        if self.generator is not None:
+            seed = self.epoch + self.generator.initial_seed()
+        else:
+            seed = self.epoch
+        g.manual_seed(seed)
+        n = len(self.data_source)
```
Comment (BenjaminBossan) [inline src/accelerate/data_loader.py:89]: The implementation starting from here is quite different from the original `RandomSampler.__iter__`. Why is that?

File: src/accelerate/data_loader.py [hunk 2]
```diff
@@ -412,8 +460,19 @@ def __iter__(self):
...
         self.end()
 
+    def set_epoch(self, epoch: int):
+        # In case it is manually passed in, the user can set it to what they like
+        if self.iteration != epoch:
+            self.iteration = epoch
+        if hasattr(self.batch_sampler, "sampler") and hasattr(self.batch_sampler.sampler, "set_epoch"):
+            self.batch_sampler.sampler.set_epoch(epoch)
+        elif hasattr(self.dataset, "set_epoch"):
+            self.dataset.set_epoch(epoch)
```
Comment (BenjaminBossan) [inline src/accelerate/data_loader.py:474]: Can we get a comment here about what types of `Dataset`s would have that method?

**Review by BenjaminBossan — COMMENTED** [review]
Wow, this was a tough nut to crack. This looks good and should hopefully enable future runs to be completely reproducible.

I have some comments, nothing big, please take a look. Also, there are several typos concerning "reproduc**a**bility" => "reproduc**i**bility"

File: src/accelerate/data_loader.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/accelerate/data_loader.py:89]: Oh I see, I haven't checked different versions. It looks like the change to using `yield from` is already in PyTorch since v1.7.0 though:

https://github.com/pytorch/pytorch/blame/v1.7.0/torch/utils/data/sampler.py

I checked the PR where this was changed:

https://github.com/pytorch/pytorch/pull/40026

Honestly, not sure if we care so much about it. However, I think the changes from this PR would be worthwhile having here as well:

https://github.com/pytorch/pytorch/pull/103339

which would be an argument in favor of copying the current implementation. But I'm fine with either way.

(Too bad that PyTorch has not moved the seeding to a separate method, which would allow us to override it and leave `__iter__` as is. With the given implementation, I don't think we can seed and call `super().__iter__()`.)

**Review by BenjaminBossan — APPROVED** [review]
I dug a bit deeper on the `__iter__` implementation of `RandomSampler`. Please take a look. Regardless of that, I'm fine with the PR, feel free to merge if you disagree with my comment.

File: src/accelerate/data_loader.py (hunk 1 above)
Comment (muellerzr) [reply src/accelerate/data_loader.py:89]: Fair point, brought it back in. We sadly can't just call `super().__iter__()`, so copy/pasted with a reference of what is going on/where

**Review by LysandreJik — APPROVED** [review]
Thank you!

File: src/accelerate/data_loader.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/accelerate/data_loader.py:89]: Note that this still does not contain the change from https://github.com/pytorch/pytorch/pull/103339, which uses `map(int, ...)` instead of `tolist()`. This is to avoid eagerly creating a Python list, which is very inefficient for big `n`.
